lib/src/main/java/io/ably/lib/transport/Hosts.java

```diff
@@ -59,20 +59,38 @@ public Hosts(String primaryHost, String defaultHost, ClientOptions options) thro
         } else {
             setPrimaryHost(defaultHost);
         }
+
+        if (options.fallbackHostsUseDefault && options.fallbackHosts != null) {
+            /* TO3k7: It is never valid to configure fallbackHost and set
+             * fallbackHostsUseDefault to true */
+            throw AblyException.fromErrorInfo(new ErrorInfo("cannot set both fallbackHosts and fallbackHostsUseDefault options", 40000, 400));
+        }
         fallbackHostsUseDefault = options.fallbackHostsUseDefault;
-        if (options.fallbackHosts == null) {
-            fallbackHosts = Arrays.copyOf(Defaults.HOST_FALLBACKS, Defaults.HOST_FALLBACKS.length);
-            fallbackHostsIsDefault = true;
+        if (primaryHostIsDefault) {
+            if (options.fallbackHosts == null && options.port == 0 && options.tlsPort == 0) {
+                if (!fallbackHostsUseDefault && options.environment != null && !options.environment.equalsIgnoreCase("production")) {
```

Comment:
Given this is inside a block where `primaryHostIsDefault` is `true`, I think this code is redundant (since if the primary host is the default, then `options.environment` is guaranteed not to be set)

Reply:
As per code, there can be a case where ```primaryHostIsDefault``` is ```true``` if non-production env is set.

Reply:
But I agree ```primaryHostIsDefault``` is always false when ```non-production``` env is set

Reply:
@vzhikserg I think you should move else block under ```if (primaryHost != null) {``` and remove this redundant check

Reply:
Idk. I feel there's still a good scope for refactoring to keep things simple. We can separate business logic for calculating a primary host and fallbacks for the respective hosts.